Framework/CurveFitting/src/Functions/Meier.cpp

```diff
@@ -0,0 +1,184 @@
+                              positiveLambda, J2);
+
+  std::valarray<double> Px;
+  calculatePx(Px, xValArray, sinAlphaSquared, cosAlphaSquared, negativeLambda, positiveLambda, J2);
+
+  std::valarray<double> Pz;
+  calculatePz(Pz, xValArray, sin2AlphaSquared, cos2AlphaSquared, negativeLambda, positiveLambda, J2);
+
+  const std::valarray<double> outValArray = A0 * gau * Lor * (1. / 3.) * (2 * Px + Pz);
+  std::copy(begin(outValArray), end(outValArray), out);
+}
+
+/**
+ * Precomputes intermediate terms used to calculate the polrization in the x and z directions. All value arrays will be
+ * resized to J2 + 2 and set to their respective quantities
+ * @param sinAlphaSquared :: sin of alpha squared
+ * @param cosAlphaSquared :: cos of alpha squared
+ * @param sin2AlphaSquared :: sin of 2*alpha squared
+ * @param cos2AlphaSquared :: cos of 2*alpha squared
+ * @param positiveLambda :: negative lambda
+ * @param negativeLambda :: positive lambda
+ * @param J2 :: 2 * total angular momentum quantum number
+ */
+void Meier::precomputeIntermediateSteps(std::valarray<double> &sinAlphaSquared, std::valarray<double> &cosAlphaSquared,
+                                        std::valarray<double> &sin2AlphaSquared,
+                                        std::valarray<double> &cos2AlphaSquared, std::valarray<double> &negativeLambda,
+                                        std::valarray<double> &positiveLambda, const double &J2) const {
+  const double FreqD = getParameter("FreqD");
+  const double FreqQ = getParameter("FreqQ");
+  const double J = J2 / 2;
+
+  const double OmegaD = 2 * M_PI * FreqD;
+  const double OmegaQ = 2 * M_PI * FreqQ;
+
+  const size_t size = int(J2 + 2);
+  sinAlphaSquared.resize(size);
+  cosAlphaSquared.resize(size);
+  sin2AlphaSquared.resize(size);
+  cos2AlphaSquared.resize(size);
+  negativeLambda.resize(size);
+  positiveLambda.resize(size);
+  std::valarray<double> Wm(size);
+
+  for (size_t i = 0; i < size; i++) {
+    const double m = static_cast<double>(i) - J;
+    const double q1 = (OmegaQ + OmegaD) * (2 * m - 1);
+    const double q2 = OmegaD * std::sqrt(J * (J + 1) - m * (m - 1));
+    const double qq = pow(q1, 2) + pow(q2, 2);
+    const double q3 = OmegaQ * (2 * pow(m, 2) - 2 * m + 1) + OmegaD;
+
+    Wm[i] = std::sqrt(qq);
+
+    if (static_cast<double>(i) < (J2 + 1)) {
+      positiveLambda[i] = 0.5 * (q3 + Wm[i]);
+    } else {
+      positiveLambda[i] = OmegaQ * pow(J, 2) - OmegaD * J;
+    }
+
+    if (i > 0) {
+      negativeLambda[i] = 0.5 * (q3 - Wm[i]);
+    } else {
+      negativeLambda[i] = OmegaQ * pow(J, 2) - OmegaD * J;
+    }
+
+    if (qq > 0) {
+      cos2AlphaSquared[i] = pow(q1, 2) / qq;
+    } else {
+      cos2AlphaSquared[i] = 0;
+    }
+    sin2AlphaSquared[i] = 1 - cos2AlphaSquared[i];
+
+    cosAlphaSquared[i] = 0.5 * (1 + std::sqrt(cos2AlphaSquared[i]));
+    sinAlphaSquared[i] = 1 - cosAlphaSquared[i];
+  }
```

Comment:
could we do something like (this is bad psedu code and can be improved):

```
Pair<Double, double> cos2AlphaSq(double q1, double q2){
If (q1 >0 || q2 >0){
Double qsq = (pow(q1,2) + pow(q2,2)
Return pow(q1,2)/qsq;
}

double getPositiveLambda(int I, double j2, Wm, omegaQ, omegaD, j, m){
    const double q3 = OmegaQ * (2 * pow(m, 2) - 2 * m + 1) + OmegaD;   
 if (static_cast<double>(i) < (J2 + 1)) {
      return  0.5 * (q3 + Wm);
    } else {
      return OmegaQ * pow(J, 2) - OmegaD * J;
    }

}

Size_t i = 0;
    const double m = static_cast<double>(i) - J;
    const double q1 = (OmegaQ + OmegaD) * (2 * m - 1);
    const double q2 = OmegaD * std::sqrt(J * (J + 1) - m * (m - 1));
cos2AlphaSquared[i], qq = cos2AlphaSq(q1,q2)
Wm[i] = std::sqrt(qq);
positiveLambda[i] = getPositiveLambda(i,  j2, Wm[i], omegaQ, omegaD, j, m)
negativeLambda[i] = 0.5 * (q3 - Wm[i]);
sin2AlphaSquared[i] = 1 - cos2AlphaSquared[i];
    cosAlphaSquared[i] = 0.5 * (1 + std::sqrt(cos2AlphaSquared[i]));
    sinAlphaSquared[i] = 1 - cosAlphaSquared[i];
 

  for (size_t i = 1; i < size; i++) {
    m = static_cast<double>(i) - J;
    q1 = (OmegaQ + OmegaD) * (2 * m - 1);
    q2 = OmegaD * std::sqrt(J * (J + 1) - m * (m - 1));


cos2AlphaSquared[i], qq = cos2AlphaSq(q1,q2)
Wm[i] = std::sqrt(qq);
positiveLambda[i] = getPositiveLambda(i,  j2, Wm[i], omegaQ, omegaD, j, m)
 negativeLambda[i] = OmegaQ * pow(J, 2) - OmegaD * J;

sin2AlphaSquared[i] = 1 - cos2AlphaSquared[i];

    cosAlphaSquared[i] = 0.5 * (1 + std::sqrt(cos2AlphaSquared[i]));
    sinAlphaSquared[i] = 1 - cosAlphaSquared[i];
  }

```

Reply:
Actually, lambda calculation can be combined into one function because of the sign. Then, I can calculate cos2alphasquared in a separate function the code will be:
```

for (size_t i = 0; i < size; i++) {
    const double m = static_cast<double>(i) - J;
    const double q1 = (OmegaQ + OmegaD) * (2 * m - 1);
    const double q2 = OmegaD * std::sqrt(J * (J + 1) - m * (m - 1));
    const double qq = pow(q1, 2) + pow(q2, 2);
    const double q3 = OmegaQ * (2 * pow(m, 2) - 2 * m + 1) + OmegaD;
    Wm[i] = std::sqrt(qq);

    positiveLambda[i] = calculateLambda(i, J2, q3, Wm[i], OmegaQ, OmegaD, true); //last parameter is a flag used to indicate if it is positive or negative
    negativeLambda[i] = calculateLambda(i, J2, q3, Wm[i], OmegaQ, OmegaD, false);
    cos2AlphaSquared[i] =  calculateCos2AlphaSquared(q1, qq);
    cosAlphaSquared[i] = 0.5 * (1 + std::sqrt(cos2AlphaSquared[i]));
    sinAlphaSquared[i] = 1 - cosAlphaSquared[i];
  }

```